demo/runtime/src/runtime/democracy.rs

```diff
@@ -151,72 +184,92 @@ pub fn next_free_ref_index() -> ReferendumIndex {
 	storage::get_or_default(REFERENDUM_COUNT)
 }
 
-pub mod public {
-	use super::*;
+pub type BoxedProposal = Box<Proposal>;
 
+impl_dispatch! {
+	pub mod public;
+	fn propose(proposal: Box<Proposal>, value: Balance) = 0;
+	fn second(proposal: PropIndex) = 1;
+	fn vote(ref_index: ReferendumIndex, approve_proposal: bool) = 2;
+}
+
+impl<'a> public::Dispatch for PublicPass<'a> {
 	/// Propose a sensitive action to be taken.
-	pub fn propose(signed: &AccountId, proposal: &Proposal, value: Balance) {
+	fn propose(self, proposal: Box<Proposal>, value: Balance) {
 		assert!(value >= minimum_deposit());
-		assert!(staking::internal::deduct_unbonded(signed, value));
+		assert!(staking::internal::deduct_unbonded(&self, value));
 
 		let index: PropIndex = storage::get_or_default(PUBLIC_PROP_COUNT);
 		storage::put(PUBLIC_PROP_COUNT, &(index + 1));
-		storage::put(&index.to_keyed_vec(DEPOSIT_OF), &(value, vec![*signed]));
+		storage::put(&index.to_keyed_vec(DEPOSIT_OF), &(value, vec![*self]));
 
 		let mut props = public_props();
-		props.push((index, proposal.clone(), *signed));
+		props.push((index, (*proposal).clone(), *self));
 		storage::put(PUBLIC_PROPS, &props);
 	}
 
 	/// Propose a sensitive action to be taken.
-	pub fn second(signed: &AccountId, proposal: PropIndex) {
+	fn second(self, proposal: PropIndex) {
 		let key = proposal.to_keyed_vec(DEPOSIT_OF);
 		let mut deposit: (Balance, Vec<AccountId>) =
 			storage::get(&key).expect("can only second an existing proposal");
-		assert!(staking::internal::deduct_unbonded(signed, deposit.0));
+		assert!(staking::internal::deduct_unbonded(&self, deposit.0));
 
-		deposit.1.push(*signed);
+		deposit.1.push(*self);
 		storage::put(&key, &deposit);
 	}
 
 	/// Vote in a referendum. If `approve_proposal` is true, the vote is to enact the proposal;
 	/// false would be a vote to keep the status quo..
-	pub fn vote(signed: &AccountId, ref_index: ReferendumIndex, approve_proposal: bool) {
+	fn vote(self, ref_index: ReferendumIndex, approve_proposal: bool) {
 		if !is_active_referendum(ref_index) {
 			panic!("vote given for invalid referendum.")
 		}
-		if staking::balance(signed) == 0 {
+		if staking::balance(&self) == 0 {
 			panic!("transactor must have balance to signal approval.");
 		}
-		let key = (*signed, ref_index).to_keyed_vec(VOTE_OF);
+		let key = (*self, ref_index).to_keyed_vec(VOTE_OF);
 		if !storage::exists(&key) {
 			let mut voters = voters_for(ref_index);
-			voters.push(signed.clone());
+			voters.push(self.clone());
 			storage::put(&ref_index.to_keyed_vec(VOTERS_FOR), &voters);
 		}
 		storage::put(&key, &approve_proposal);
 	}
 }
 
-pub mod privileged {
-	use super::*;
+impl_dispatch! {
+	pub mod privileged;
+	fn start_referendum(proposal: Box<Proposal>, vote_threshold: VoteThreshold) = 0;
```

Comment:
why does this accept a `Box`?

Reply:
because it's inside `impl_dispatch` so it needs to be pass-by-value (since the same types will be used in the `enum` item prototypes).

Reply:
just to clarify: there can be a referendum to dispatch any kind of private call? (which we've renamed to `Proposal` within this module -- I found that confusing)

Reply:
correct, though it's not "private-only", it's "privileged-only".